aea/helpers/dialogue/base.py

```diff
@@ -276,77 +328,185 @@ def is_empty(self) -> bool:
 
-        :param message: a message to be added
-        :return: True if message successfully added, false otherwise
+        These rules are designed to be fundamental to all dialogues, and enforce the following:
+
+         - message ids are consistent
+         - targets are consistent
+         - message targets are according to the reply structure of performatives
+
+        :param message: the message to be validated
+        :return: True if valid, False otherwise.
         """
-        is_extendable = self.is_valid_next_message(message)
-        if is_extendable:
-            self.incoming_extend(message)
-        return is_extendable
+        message_id = cast(int, message.get("message_id"))
+        target = cast(int, message.get("target"))
+        performative = message.get("performative")
 
-    def update(self, message: "Message") -> bool:
+        if self.is_empty:
+            result = (
+                message_id == Dialogue.STARTING_MESSAGE_ID
+                and target == Dialogue.STARTING_TARGET
+                and performative == self.initial_performative()
+            )
+        else:
+            last_message_id = self.last_message.message_id  # type: ignore
+            target_message = self.get_message(target)
+            if target_message is not None:
+                target_performative = target_message.get("performative")
+                result = (
+                    message_id == last_message_id + 1
+                    and 1 <= target <= last_message_id
+                    and performative in self.reply(target_performative)
+                )
+            else:
+                result = False
+        return result
+
+    def _additional_rules(self, message: Message) -> bool:
         """
-        Extend the list of incoming/outgoing messages with 'message', if 'message' is valid
+        Check whether 'message' is a valid next message in the dialogue, according to additional rules.
 
-        :param message: a message to be added
-        :return: True if message successfully added, false otherwise
+        These rules are designed to be less fundamental than basic rules and subject to change.
+        Currently the following is enforced:
+
+         - A message targets the message strictly before it in the dialogue
+
+        :param message: the message to be validated
+        :return: True if valid, False otherwise.
         """
-        if message.is_incoming:
-            is_successful = self.incoming_safe_extend(message)
+        if self.is_empty:
+            result = True
         else:
-            is_successful = self.outgoing_safe_extend(message)
-        return is_successful
+            target = cast(int, message.get("target"))
+            last_target = self.last_message.target  # type: ignore
+            result = target == last_target + 1
+        return result
 
     @abstractmethod
-    def is_valid_next_message(self, message: "Message") -> bool:
+    def initial_performative(self):
```

Comment:
This means we need an abstract performative class to specify the signature correctly...